core/vm/evm.go

```diff
@@ -435,6 +443,33 @@ func (evm *EVM) create(caller ContractRef, codeAndHash *codeAndHash, gas uint64,
 	if !evm.Context.CanTransfer(evm.StateDB, caller.Address(), value) {
 		return nil, common.Address{}, gas, ErrInsufficientBalance
 	}
+
+	// Initialise a new contract and set the code that is to be used by the EVM.
+	// The contract is a scoped environment for this execution context only. If
+	// the initcode is EOF, contract.Container will be set.
+	contract := NewContract(caller, AccountRef(address), value, gas)
+	contract.SetCodeOptionalHash(&address, codeAndHash)
+	contract.IsDeployment = true
+
+	// Validate initcode per EOF rules. If caller is EOF and initcode is legacy, fail.
+	isInitcodeEOF := hasEOFMagic(codeAndHash.code)
+	if isInitcodeEOF {
```

Comment:
We could think about moving this a bit later. Right now we do the heavy things (unmarshalling and validation) before we actually deduct the gas and check for nonce overflow

Reply:
This piece of goes away when [EIP-7873](https://eips.ethereum.org/EIPS/eip-7873) arrives and we remove [EIP-7698](https://eips.ethereum.org/EIPS/eip-7698), so I wouldn't spin too many cycles optimizing it.